Add suggested changes from API guild discussion (POST usage) (#791)
zalando/restful-api-guidelines: chapters/http-status-codes-and-errors.adoc

[[redirection-codes]]
=== Redirection codes

[.do-not-use]
[[status-code-301]]
==== 301 Moved Permanently {rfc-status-301} {do-not-use} {ALL}
[.indent]
This and all future requests should be directed to the given URI. See also

> <<251>>.

Reviewer: The link to rule `250` is surely wrong, but I still think, that adding the reference to `251` on the top is shorter and more effective than adding the reference to each effected `3xx` response code. The exception of `304` in my opinion is marked sufficiently be the rule.

[.do-not-use]
[[status-code-302]]
==== 302 Found {rfc-status-302} {do-not-use} {ALL}
[.indent]
This is a temporary redirect where some clients {MAY} change the request method
from {POST} to {GET}. Mainly used for dismissing and redirecting form
submissions in browsers. See also <<251>>.

[.do-not-use]
[[status-code-303]]
==== 303 See Other {rfc-status-303} {do-not-use} {POST} {PUT} {PATCH} {DELETE}
[.indent]
The response to the request can be found under another URI using a {GET}
method. A disambiguated version of {302} for the case where the client {MUST}
change the method to {GET}. See also <<251>>.

[[status-code-304]]
==== 304 Not Modified {rfc-status-304} {document} {GET} {HEAD}
[.indent]

> Indicates that a conditional {GET} or {HEAD} request would have resulted in

Reviewer: ```suggestion
Indicates that a conditional {GET} or {HEAD} request would have resulted in a
```

[.indent]
The client is not abiding by the rate limits in place and has sent too many
requests (see <<153>>).

[[status-code-431]]
==== 431 Request Header Fields Too Large {rfc-status-431} {do-not-document} {ALL}
[.indent]
The server is not able to process the request because the request headers are
too large. Usually used by gateways and proxies with memory limits.

=== Server side error codes

[[status-code-500]]
==== 500 Internal Server Error {rfc-status-500} {use} {do-not-document} {ALL}
[.indent]
A generic error indication for an unexpected server execution problem. Clients
should be careful with retrying on this response, since the nature of the
problem is unknown and must be expected to continue.

[[status-code-501]]
==== 501 Not Implemented {rfc-status-501} {document} {ALL}
[.indent]
Server cannot fulfill the request, since the endpoint is not implemented yet.
Usually this implies future availability, but retrying now is not recommended.
May be documented on endpoints that are planned to be implemented in the
future to indicate that they are still not available.

[[status-code-502]]
==== 502 Bad Gateway {rfc-status-502} {ALL}

> [.indent]is meaningless

Reviewer: This looks like an unintended text fragment?

```suggestion
[.indent]
```

The server, while acting as a gateway or proxy, received an invalid response
from an inbound server attempting to fulfill the request. May be used by a
server to indicate that an inbound service is creating an unexpected result
instead of {500}. Clients should be careful with retrying on this response,
since the nature of the problem is unknown and must be expected to continue.

[[status-code-503]]
==== 503 Service Unavailable {rfc-status-503} {use} {do-not-document} {ALL}
[.indent]
Service is (temporarily) not available, e.g. if a required component or inbound
service is not available. Client are encouraged to retry requests following an
exponential back off pattern. If possible, the service should indicate how long
the client should wait by setting the {Retry-After} header.

[[status-code-504]]
==== 504 Gateway Timeout {rfc-status-504} {use} {ALL}
[.indent]
The server, while acting as a gateway or proxy, did not receive a timely
response. May be used by servers to indicate that an inbound service cannot
process the request fast enough. Client may retry the request immediately

> exactly once, to check whether warming up the service solved the problem.

Reviewer: `A client` or `Clients` I think.

[.do-not-use]
[[status-code-505]]
==== 505 HTTP Version Not Supported {rfc-status-505} {do-not-use} {ALL}
[.indent]
The server does not support the HTTP protocol version used in the request.

> Technical response code that serves not use case in RESTful APIs.

Reviewer: Copilot is needs some polishing:

```suggestion
A technical response code that serves no use case in RESTful APIs.
```
